FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageJobQueueClient.cs

```diff
@@ -187,58 +192,71 @@ await foreach (var pageResult in jobEntityQueryResult.AsPages().WithCancellation
-                        new TableTransactionAction(TableTransactionActionType.UpdateReplace, jobLockEntity, jobLockEntity.ETag));
+                    foreach (var jobLockEntity in jobLockEntities.Where(jobLockEntity => !jobLockEntity.ContainsKey(JobLockEntityProperties.JobMessageId)))
+                    {
+                        var response = await _azureJobMessageQueueClient.SendMessageAsync(
+                            new JobMessage(jobLockEntity.PartitionKey, jobLockEntity.GetString(JobLockEntityProperties.JobInfoEntityRowKey), jobLockEntity.RowKey).ToString(),
+                            cancellationToken);
 
-                    _ = await _azureJobInfoTableClient.SubmitTransactionAsync(transactionUpdateActions, cancellationToken);
-                }
-                catch (RequestFailedException ex) when (IsSpecifiedErrorCode(ex, AzureStorageErrorCode.UpdateEntityPreconditionFailedErrorCode))
-                {
-                    _logger.LogWarning(ex, "Update job info entity conflicts.");
+                        jobLockEntity[JobLockEntityProperties.JobMessagePopReceipt] = response.Value.PopReceipt;
+                        jobLockEntity[JobLockEntityProperties.JobMessageId] = response.Value.MessageId;
+                    }
+
+                    // step 7: update message id and message pop receipt to job lock entity
+                    // if enqueue concurrently, it is possible that
+                    // 1. one job sends message and updates entity, another job do nothing
+                    // 2. two jobs both send message, while only one job update entity successfully
+                    try
+                    {
+                        var transactionUpdateActions = jobLockEntities.Select(jobLockEntity =>
+                            new TableTransactionAction(TableTransactionActionType.UpdateReplace, jobLockEntity, jobLockEntity.ETag));
+
+                        _ = await _azureJobInfoTableClient.SubmitTransactionAsync(transactionUpdateActions, cancellationToken);
+                    }
+                    catch (RequestFailedException ex) when (IsSpecifiedErrorCode(ex, AzureStorageErrorCode.UpdateEntityPreconditionFailedErrorCode))
+                    {
+                        _logger.LogWarning(ex, "Update job info entity conflicts.");
+                    }
                 }
-            }
 
-            jobInfos = jobInfoEntities.Select(entity => entity.ToJobInfo<TJobInfo>()).ToList();
+                jobInfos = jobInfoEntities.Select(entity => entity.ToJobInfo<TJobInfo>()).ToList();
 
-            _logger.LogInformation($"Enqueue jobs '{string.Join(",", jobInfos.Select(jobInfo => jobInfo.Id).ToList())}' successfully.");
+                _logger.LogInformation($"Enqueue jobs '{string.Join(",", jobInfos.Select(jobInfo => jobInfo.Id).ToList())}' successfully.");
 
-            return jobInfos;
+                return jobInfos;
+            }
+            catch (Exception ex)
```

Comment:
typo in log information for enqueue successfully. `$"Enqueue jobs {string.Join(",", jobInfos.Select(jobInfo => jobInfo.Id).ToList())} successfully."`